Added ForbiddenApis Plugin and performed possible optimizations

kevinfwl opened the pull request:
Added the ForbiddenApis plugin that detects a list of possibly inefficient methods and fails CI. Currently, this list only includes string manipulation methods that use regexes and `Class.forName()`. @richardstartin let me know if I should add more to the filter list. I also replaced any inefficient methods with a possibly more optimized variation by not using regexes internally.

pmbauer commented:
> and fails CI
I also saw this check was removed from tests.  Was it too noisy or expensive?
We should provide developers local feedback as early as possible before pushing to CI, can easily waste 10 minutes of cycle time if the first they know there is a violation is CI.  It's likely they run tests before pushing.

pmbauer commented:
> and fails CI

I also saw this check was removed from tests.  Was it too noisy or expensive?
We should provide developers local feedback as early as possible before pushing to CI, can easily waste 10 minutes of cycle time if the first they know there is a violation is CI.  It's likely they run tests before pushing.

kevinfwl commented:
> > and fails CI
> 
> I also saw this check was removed from tests ([06e08fa](https://github.com/DataDog/dd-trace-java/commit/06e08fa1d89a8980f49ba53f80c0267049f77231)) Was it too noisy or expensive?
> We should provide developers local feedback as early as possible before pushing to CI, can easily waste 10 minutes of cycle time if the first they know there is a violation is CI. It's likely they run tests before pushing.

For the removal of checking tests, the idea is to reduce cost. Since this is more of a code optimization check, I was thinking it is not necessary unless for production code. To the second point, running the task `./gradlew forbiddenApis` will give developers feedback of any failing checks. Is that what you are looking for?

richardstartin reviewed (commented):
I think we need to talk about what we want to achieve here. 

I think the goal should be the identification of code smells which point to potential performance problems, which can then be investigated in isolation without running thousands of benchmark applications. 

I don't think this will work if we find, for example, `String.split` and replace it with our own, probably slightly worse `String.split` implementation - we should be asking why are we creating a `String[]` at all in each particular case. The mitigation for some of the smells we identify may not be local; some may require a larger refactoring. One of the risks in subverting the checks by simply emulating a call to a forbidden API is that we will no longer be able to identify it like this, and will need exhaustive measurement to find them, so we should avoid doing this. 

This approach will produce a low enough volume of code smells that we can reasonably expect to benchmark each forbidden API against its candidate replacement, to validate the change before accepting it. If you like, I can help you set up (and more importantly, sanity check) some JMH benchmarks for some of these changes.

pmbauer commented:
Not entirely. A separate gradle task will be forgotten and violations won’t
get discovered until CI. We have a number of other CI-only checks that make
CI a frustrating gate due to the poor cycle time there.

Having said that, if the check is slowing tests down significantly, I agree
with the pragmatic decision.

On Sunday, February 28, 2021, Kevin Feng <notifications@github.com> wrote:

> and fails CI
>
> I also saw this check was removed from tests (06e08fa
> <https://github.com/DataDog/dd-trace-java/commit/06e08fa1d89a8980f49ba53f80c0267049f77231>)
> Was it too noisy or expensive?
> We should provide developers local feedback as early as possible before
> pushing to CI, can easily waste 10 minutes of cycle time if the first they
> know there is a violation is CI. It's likely they run tests before pushing.
>
> For the removal of checking tests, the idea is to reduce cost. Since this
> is more of a code optimization check, I was thinking it is not necessary
> unless for production code. To the second point, running the task ./gradlew
> forbiddenApis will give developers feedback of any failing checks. Is
> that what you are looking for?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/DataDog/dd-trace-java/pull/2449#issuecomment-787537077>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADOXTBQGYQ74UJ4NSCHQ73TBLAPLANCNFSM4YI7WKOA>
> .
>

richardstartin commented on dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/jdbc/JDBCConnectionUrlParser.java:
```diff
@@ -60,59 +61,63 @@
 
-      type = jdbcUrl.substring(0, hostIndex);
+      final String type = jdbcUrl.substring(0, hostIndex);
+      final String urlPart1;
```
this looks better now

richardstartin commented on dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/jdbc/JDBCConnectionUrlParser.java:
```diff
@@ -740,7 +746,7 @@ public static DBInfo parse(String connectionUrl, final Properties props) {
     final String[] pairs = query.split(separator);
     for (final String pair : pairs) {
```
I think this would be a nice target for a for loop, but I'm not too worried about this code because we do have benchmark coverage for it, and this just doesn't execute very often

kevinfwl replied:
I added a commit for this, please let me know if it is something that we are looking for and should keep. Would be okay to revert it otherwise 👍

richardstartin commented on dd-java-agent/agent-jmxfetch/src/main/java/datadog/trace/agent/jmxfetch/JMXFetch.java:
```diff
@@ -152,7 +152,13 @@ private static String getReporter(final Config config) {
       while (scanner.hasNext()) {
         String config = scanner.next();
         integrationName.clear();
-        integrationName.add(config.replace(".yaml", ""));
```
Go for readability here, it just doesn't execute enough to be worth optimising

richardstartin commented on dd-java-agent/instrumentation/spymemcached-2.12/src/main/java/datadog/trace/instrumentation/spymemcached/MemcacheClientDecorator.java:
```diff
@@ -54,17 +54,13 @@ protected String dbHostname(MemcachedConnection connection) {
+    StringBuilder builder = new StringBuilder(methodName);
+    if (builder.indexOf("async") == 0) builder.delete(0, "async".length());
+    if (builder.indexOf("CAS") == 0) builder.replace(0, "CAS".length(), "cas");
+    builder.replace(0, 1, String.valueOf(Character.toLowerCase(builder.charAt(0))));
 
-    // Lowercase first letter
-    chars[0] = Character.toLowerCase(chars[0]);
-
-    span.setResourceName(new String(chars));
+    span.setResourceName(builder.toString());
```
this seems reasonable but let's benchmark it to check

richardstartin replied:
Also try using a precompiled pattern

richardstartin commented on dd-java-agent/instrumentation/couchbase-2.0/src/main/java/datadog/trace/instrumentation/couchbase/client/CouchbaseOnSubscribe.java:
```diff
@@ -16,8 +16,14 @@ public CouchbaseOnSubscribe(
     final Class<?> declaringClass = method.getDeclaringClass();
-    final String className =
-        declaringClass.getSimpleName().replace("CouchbaseAsync", "").replace("DefaultAsync", "");
+    StringBuilder builder = new StringBuilder(declaringClass.getSimpleName());
```
Do you want to use the `Strings.replace` utility here?

Another thing you could do, which might work out better, is as follows:

```
private static final Pattern REPLACEMENTS = Pattern.compile("CouchbaseAsync| DefaultAsync");
...
String resourceName = REPLACEMENTS.matcher(method.getDeclaringClass().getSimpleName()).replaceAll("");
```

but let's benchmark it in isolation to find out before accepting the change.

tylerbenson replied:
This could potentially be optimized further by storing the result in a `ClassValue`.

richardstartin commented on dd-java-agent/instrumentation/aws-java-sdk-1.11.0/src/main/java/datadog/trace/instrumentation/aws/v0/AwsSdkClientDecorator.java:
```diff
@@ -22,15 +23,15 @@
             @Override
             public String apply(Class<?> input) {
-              return input.getSimpleName().replace("Request", "");
+              return Strings.replace(input.getSimpleName(), "Request", "");
```
I'm tempted to say we should just suppress these since they're cached calls anyway, and this method only compiles a `Pattern` on JDK8 or earlier.

richardstartin commented on dd-java-agent/instrumentation/jms/src/main/java/datadog/trace/instrumentation/jms/MessageExtractAdapter.java:
```diff
@@ -16,10 +16,13 @@
-              // true story \/
-              .replace("__dash__", "-")
-              .toLowerCase();
+
+          StringBuilder builder = new StringBuilder(key);
+          int i;
+          while ((i = builder.indexOf("$")) != -1) builder.replace(i, i + "$".length(), "-");
+          while ((i = builder.indexOf("__dash__")) != -1)
+            builder.replace(i, i + "__dash__".length(), "-");
+          return builder.toString().toLowerCase();
```
Let's drop this change since removing the "__dash__" thing has been planned for a long time, and it's just a question of merging the other PR now.

kevinfwl commented:
> I think we need to talk about what we want to achieve here.
> 
> I think the goal should be the identification of code smells which point to potential performance problems, which can then be investigated in isolation without running thousands of benchmark applications.
> 
> I don't think this will work if we find, for example, `String.split` and replace it with our own, probably slightly worse `String.split` implementation - we should be asking why are we creating a `String[]` at all in each particular case. The mitigation for some of the smells we identify may not be local; some may require a larger refactoring. One of the risks in subverting the checks by simply emulating a call to a forbidden API is that we will no longer be able to identify them this way, and will need exhaustive measurement to find them, so we should avoid doing this.
> 
> This approach will produce a low enough volume of code smells that we can reasonably expect to benchmark each forbidden API against its candidate replacement, to validate the change before accepting it. If you like, I can help you set up (and more importantly, sanity check) some JMH benchmarks for some of these changes.

Yeah, I also thinking if there would actually be any visible performance improvements with these changes. A benchmark would make a lot of sense.

kevinfwl commented:
> Not entirely. A separate gradle task will be forgotten and violations won’t get discovered until CI. We have a number of other CI-only checks that make CI a frustrating gate due to the poor cycle time there. Having said that, if the check is slowing tests down significantly, I agree with the pragmatic decision.
> […](#)
> On Sunday, February 28, 2021, Kevin Feng ***@***.***> wrote: and fails CI I also saw this check was removed from tests ([06e08fa](https://github.com/DataDog/dd-trace-java/commit/06e08fa1d89a8980f49ba53f80c0267049f77231) <[06e08fa](https://github.com/DataDog/dd-trace-java/commit/06e08fa1d89a8980f49ba53f80c0267049f77231)>) Was it too noisy or expensive? We should provide developers local feedback as early as possible before pushing to CI, can easily waste 10 minutes of cycle time if the first they know there is a violation is CI. It's likely they run tests before pushing. For the removal of checking tests, the idea is to reduce cost. Since this is more of a code optimization check, I was thinking it is not necessary unless for production code. To the second point, running the task ./gradlew forbiddenApis will give developers feedback of any failing checks. Is that what you are looking for? — You are receiving this because your review was requested. Reply to this email directly, view it on GitHub <[#2449 (comment)](https://github.com/DataDog/dd-trace-java/pull/2449#issuecomment-787537077)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AADOXTBQGYQ74UJ4NSCHQ73TBLAPLANCNFSM4YI7WKOA> .

ah I see, in that case, what would be an ideal way to go about this?

tylerbenson commented on dd-java-agent/instrumentation/twilio/src/main/java/datadog/trace/instrumentation/twilio/TwilioClientDecorator.java:
```diff
@@ -47,7 +47,7 @@ public AgentSpan onServiceExecution(
     // Drop common package prefix (com.twilio.rest)
     final String simpleClassName =
-        serviceExecutor.getClass().getCanonicalName().replaceFirst("^com\\.twilio\\.rest\\.", "");
+        serviceExecutor.getClass().getCanonicalName().substring("com.twilio.rest.".length());
```
Something else that might be better accessed via a `ClassValue`

tylerbenson commented on gradle/forbiddenApiFilters/main.txt:
```diff
@@ -0,0 +1,9 @@
+java.lang.Class#forName(java.lang.String)
```
Please add a comment explaining why to exclude this one.

tylerbenson reviewed (approved):
nice optimizations!  good work.

richardstartin commented on dd-java-agent/instrumentation/jms/src/main/java/datadog/trace/instrumentation/jms/MessageInjectAdapter.java:
```diff
@@ -12,7 +13,7 @@
   @Override
   public void set(final Message carrier, final String key, final String value) {
-    final String propName = key.replace("-", "__dash__");
+    final String propName = Strings.replace(key, "-", "__dash__");
```
This will be solved in another PR

richardstartin commented on internal-api/src/main/java/datadog/trace/util/Strings.java:
```diff
@@ -25,4 +25,22 @@ public static String join(CharSequence joiner, CharSequence... strings) {
     return "";
   }
+
+  // reimplementation of string functions without regex
+  public static String replace(String str, String delimiter, String replacement) {
+    StringBuilder sb = new StringBuilder(str);
+    int matchIndex, curIndex = 0;
+    while ((matchIndex = sb.indexOf(delimiter, curIndex)) != -1) {
+      sb.replace(matchIndex, matchIndex + delimiter.length(), replacement);
+      curIndex = matchIndex + replacement.length();
+    }
+    return sb.toString();
+  }
+
+  public static String replaceFirst(String str, String delimiter, String replacement) {
+    StringBuilder sb = new StringBuilder(str);
+    int i = sb.indexOf(delimiter);
+    if (i != -1) sb.replace(i, i + delimiter.length(), replacement);
+    return sb.toString();
+  }
```
Would be good to benchmark these against JDK String API equivalents

richardstartin commented on dd-java-agent/instrumentation/couchbase-2.0/src/main/java/datadog/trace/instrumentation/couchbase/client/CouchbaseOnSubscribe.java:
```diff
@@ -8,17 +8,29 @@
 
 public class CouchbaseOnSubscribe extends TracedOnSubscribe {
+
+  private static final ClassValue<String> CLASS_NAME =
+      new ClassValue<String>() {
+        @Override
+        protected String computeValue(Class<?> declaringClass) {
+          StringBuilder builder = new StringBuilder(declaringClass.getSimpleName());
+          int i;
+          while ((i = builder.indexOf("CouchbaseAsync")) != -1)
+            builder.delete(i, i + "CouchbaseAsync".length());
+          while ((i = builder.indexOf("DefaultAsync")) != -1)
+            builder.delete(i, i + "DefaultAsync".length());
+          return builder.toString();
+        }
+      };
```
I think you can actually use a class we already have for this - `QualifiedClassNameCache` - which is used in the `AwsSdkClientDecorator`. Setting it up is a stupidly cumbersome:

```java
  private static final QualifiedClassNameCache NAMES = new QualifiedClassNameCache(new Function<Class<?>, CharSequence>() {
    @Override
    public String apply(Class<?> input) {
      return input.getSimpleName().replace("CouchbaseAsync", "").replace("DefaultAsync", "");
    }
  }, Functions.PrefixJoin.of(".", Functions.<CharSequence>zero()));)
```

below the call would be just `NAMES.getQualifiedName(method.getDeclaringClass(), method.getName())`.

(You'll have to add the anonymous class `CouchbaseOnSubscribe$1` to the instrumentations.)
